File: Data Structures & Algorithms/level-order-traversal-of-binary-tree/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def levelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        # BFS 
        res = []
        if root is None:
            return res
        to_explore = [root]
        next_level = []
        curr_level = []
        # Starting at root
        while to_explore != []:
            # add children to to_explore queue
            curr = to_explore.pop()
            if curr.left:
                next_level.insert(0, curr.left)
            if curr.right:
                next_level.insert(0, curr.right)
            # add root to result
            curr_level.append(curr.val)
            if to_explore == []:
                res.append(curr_level[:])
                to_explore = next_level[:]
                next_level = []
                curr_level = []
        return res

File: Data Structures & Algorithms/depth-of-binary-tree/submission-0.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    int maxDepth(TreeNode* root) {
        // base case
        if (root == nullptr) return 0;
        //if (root->left == nullptr && root->right == nullptr) return 1;

        // recursive case
        int leftDepth = maxDepth(root->left);
        int rightDepth = maxDepth(root->right);
        if (leftDepth > rightDepth) return 1 + leftDepth;
        else return 1 + rightDepth;
    }
};
